go-controller/pkg/clustermanager/zone_cluster_controller.go

```diff
@@ -0,0 +1,355 @@
+	allocatedNodeID, err := zcc.idAllocator.AllocateID(node.Name)
+	if err != nil {
+		return fmt.Errorf("failed to allocate an id to the node %s : err - %w", node.Name, err)
+	}
+	klog.V(5).Infof("Allocated id %d to the node %s", allocatedNodeID, node.Name)
+	nodeAnnotations := util.UpdateNodeIDAnnotation(nil, allocatedNodeID)
+
+	return zcc.updateNodeAnnotationWithRetry(node.Name, nodeAnnotations)
+}
+
+// handleAddUpdateNodeEvent handles the delete node event
+func (zcc *zoneClusterController) handleDeleteNode(node *corev1.Node) error {
+	zcc.idAllocator.ReleaseID(node.Name)
+	return nil
+}
+
+func (zcc *zoneClusterController) syncNodes(nodes []interface{}) error {
+	return zcc.syncNodeIDs(nodes)
+}
+
+func (zcc *zoneClusterController) syncNodeIDs(nodes []interface{}) error {
+	duplicateIdNodes := []string{}
+
+	for _, nodeObj := range nodes {
+		node, ok := nodeObj.(*corev1.Node)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNodes: %v", nodeObj)
+		}
+
+		nodeID := util.GetNodeID(node)
+		if nodeID != -1 {
+			klog.Infof("Node %s has the id %d set", node.Name, nodeID)
+			if err := zcc.idAllocator.ReserveID(node.Name, nodeID); err != nil {
+				// The id set on this node is duplicate.
+				klog.Infof("Node %s has a duplicate id %d set", node.Name, nodeID)
+				duplicateIdNodes = append(duplicateIdNodes, node.Name)
+			}
+		}
+	}
+
+	for i := range duplicateIdNodes {
+		newNodeID, err := zcc.idAllocator.AllocateID(duplicateIdNodes[i])
+		if err != nil {
+			return fmt.Errorf("failed to allocate id for node %s : err - %w", duplicateIdNodes[i], err)
+		} else {
+			klog.Infof("Allocated new id %d for node %q", newNodeID, duplicateIdNodes[i])
+		}
+	}
+
+	return nil
+}
+
+// UpdateNodeAnnotationWithRetry update node's annotation with the given node annotations.
+func (zcc *zoneClusterController) updateNodeAnnotationWithRetry(nodeName string,
+	nodeAnnotations map[string]string) error {
+	// Retry if it fails because of potential conflict which is transient. Return error in the
+	// case of other errors (say temporary API server down), and it will be taken care of by the
+	// retry mechanism.
+	resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Informer cache should not be mutated, so get a copy of the object
+		node, err := zcc.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return fmt.Errorf("unable to fetch nodes from watcher, err: %v", err)
+		}
+
+		cnode := node.DeepCopy()
+		if cnode.Annotations == nil {
+			cnode.Annotations = make(map[string]string)
+		}
+		for k, v := range nodeAnnotations {
+			cnode.Annotations[k] = v
+		}
+		return zcc.kube.UpdateNode(cnode)
+	})
+	if resultErr != nil {
+		return fmt.Errorf("failed to update node %s annotation", nodeName)
+	}
+	return nil
+}
+
+// zoneClusterControllerEventHandler object handles the events
+// from retry framework.
+type zoneClusterControllerEventHandler struct {
```

Comment:
I guess it can be considered later ?

I think this one came out of our discussion in the cluster manager patch to have separate controller for this.  But I need to dig into that PR to be sure :) 

Following commits in this PR do add more functionality in this controller.